### PR TITLE
Support per metric custom tags

### PR DESCRIPTION
**Pull request by x1a0**
Poor man's support of custom tags. This could solve https://github.com/iZettle/dropwizard-metrics-influxdb/issues/42.

### How to use

When defining key of a metric, use the pattern `name|tag1:val1|tag2:val2|...|tagN:valN` to add custom tags.

For example, given a metric key `user.created|country:SE`, it will be sent to influxdb measurement `user.created` (if no measurement mapping specified) with a tag `country` and corresponding value `SE`.

To create/retrieve the metric handler in Dropwizard:
```
metricRegistry.counter(name("user", "created|country:SE"));
```

### About grouped gauges

For grouped gauges, custom tags are the last part of keys.

For example, `gauge.a|country:SE`, `gauge.b|country:DE`, `gauge.|country:SE` and `gauge|country:SE` could be grouped.

### CODE REVIEW

**Comment by rickard-von-essen-iz** [comment]
👍

File: metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java [hunk 1]
```diff
@@ -212,29 +204,39 @@ public void report(
...
-            }
+            counters.forEach((key, counter) -> {
+                parse(key, (name, tags) -> reportCounter(name, counter, now, tags));
+            });
 
-            for (Map.Entry<String, Histogram> entry : histograms.entrySet()) {
-                reportHistogram(entry.getKey(), entry.getValue(), now);
-            }
+            histograms.forEach((key, histogram) -> {
+                if (canSkipMetric(key, histogram)) {
```
Comment (xiaodong-izettle) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: could convert this to a `filter()` here?

Comment (fredrikbackstrom) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: with these kinds of changes we are making the repo java8 dependent, has there been a decision before to not make it java8 or do we even have other parts of the code that make J8 a must?

**Comment by alorlea** [comment]
:+1: :100:

File: metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java (hunk 1 above)
Comment (xiaodong-izettle) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: Hehe, I think we have already mandated java8 in our pom file...

Comment (x1a0) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: @xiaodong-izettle Would need to do `histograms.entrySet().stream.filter(...).forEach(...)` - does it worth? (not sure memory/performance wise what happens under the hood).

@fredrikbackstrom Sorry I did not think about it. If we decide to not make java8 a must I can refactor the codes. However one question there: what's the elegant way to do the `parse()` method then as I want to 2 returns from that method: name and tags.

Comment (xiaodong-izettle) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: Hehe, probably not worthy for `histograms.entrySet().stream.filter(...).forEach(...)`, but the memory/performance impact should be neglectable... I'm OK... :-)

File: metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java [hunk 2]
```diff
@@ -246,15 +248,29 @@ public void report(
         }
     }
 
+    private void parse(String key, BiConsumer<String, Map<String, String>> report) {
+        String[] exploded = key.split("\\|");
+        Map<String, String> tags = new HashMap<>();
+        String name = exploded[0];
+        for (int i = 1; i < exploded.length; i++) {
+            String[] kv = exploded[i].split(":");
+            tags.put(kv[0], kv[1]);
```
Comment (x1a0) [inline metrics-influxdb/src/main/java/com/izettle/metrics/influxdb/InfluxDbReporter.java]: Hmm... that would need something like `Arrays.stream(exploded).skip(1).map(...)` - not sure how to do it exactly yet. But it feels a bit overkilling I guess?

**Comment by x1a0** [comment]
This solution is a bit tricky and twisted. We decided to directly push metrics data into influxdb instead of going through dropwizard metrics.
Thanks for all the reviews. I am closing this.

**Comment by xiaodong-izettle** [comment]
Yeah, that probably makes more sense... :-)

**Pull request closed without merging**